src/UCTNode.cpp

```diff
@@ -376,43 +437,58 @@ class NodeComp : public std::binary_function<UCTNode::node_ptr_t&,
 
 void UCTNode::sort_children(int color) {
     LOCK(get_mutex(), lock);
-    std::stable_sort(rbegin(m_children), rend(m_children), NodeComp(color));
+    // Because sort depends on visits there's no need to expand all nodes.
+
+    std::stable_sort(rbegin(m_expanded), rend(m_expanded), NodeComp(color));
 }
 
 UCTNode& UCTNode::get_best_root_child(int color) {
     LOCK(get_mutex(), lock);
-    assert(!m_children.empty());
 
-    return *(std::max_element(begin(m_children), end(m_children),
+    if (m_expanded.size() < 2) {
+        // NodeComp first checks visits so only need to consider m_expanded,
+        // unless few things has been expanded then for simplicity expand_all.
+        // This can be improved by expanding and returning node with max score.
+        expand_all();
```

Comment:
Not sure about the need to `expand_all` here. If m_expanded is empty, then this should just return the first child since all get_evals would be equal in that case.

Reply:
in fixing this I introduced a subtle bug which worries my, I'm backing out some of the clever optimizations and changing a number of places to call expand_all()

If we want to optimize performance later I think that's a fine thing but for now I'm more worried about making mistakes.